powerinfer-py/powerinfer/solver.py

```diff
@@ -0,0 +1,90 @@
+#!/usr/bin/env python
+# coding=utf-8
+import argparse
+from cvxopt.glpk import ilp
```

Comment:
Using convex optimization. Specifically [GLPK: GNU Linear Programming Kit.](https://en.wikipedia.org/wiki/GNU_Linear_Programming_Kit). Linear programming is an optimization paradigm for matrix optimizations of the following form: 

<img width="542" alt="image" src="https://github.com/ggerganov/llama.cpp/assets/9305875/31ca64ee-2c09-43b6-b8e7-7b750fa34565">



See [polyhedral compilation](https://www.cse.iitk.ac.in/users/swarnendu/courses/autumn2020-cs610/intro-to-polyhedral-model.pdf) for related technique in AI compilers.

Reply:
See compressive sensing: [intro to compressive sensing.pdf](https://github.com/ggerganov/llama.cpp/files/13724875/intro.to.compressive.sensing.pdf) for very influent techniques on how L1 norm minimization leads to sparsity while maintaining good data recovery.

![Uploading image.png…]()

Reply:
ILP, the specific linear programming technique used here is integer linear programming. It's basically linear programming but where the matrix elements are integers. See this:
<img width="659" alt="image" src="https://github.com/ggerganov/llama.cpp/assets/9305875/32c5eb52-ae55-4045-b055-165c48a0772b">
 https://gist.github.com/nipunbatra/7059160